### PR TITLE
Add retries on a per-specfile basis

### PR DESCRIPTION
**Pull request by bennieswart**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Add retries on a per-specfile basis, as discussed in https://github.com/webdriverio/webdriverio/issues/2484.

Previously only test-level and suite-level retries were available, which
are fine in most cases. In any test which involves state, such as on a
server or in a db, the state may be left invalid once the test fails the
first time. Any subsequent retries may have no chance of passing due to
the invalid state they would have to start with.

A new browser instance is created for each specfile, which makes this an
ideal place to hook and setup any other states (server, db). Retries on
this level would mean that the whole setup process could simply be
repeated as if it were for a new specfile, with the results of the
failing specfile (that is to be retried) being discarded.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

Documentation kindly provided by @pdirt and pending in #2614.
Will integrate docs with this PR if so requested. Otherwise, just merge both.

### CODE REVIEW

**Comment by bennieswart** [comment]
We've been running this patch for a couple of days now and the retries seem to be working nicely. I just noticed however that the reported duration of tests have dropped down to tens of millis. Only tests that have passed on their final run are correctly reporting duration. This correlates with the messages being buffered. I suspect start/end timestamps are assigned by the message handler as apposed to being contained within the message itself. Will look into this a bit later - should be an easy fix.

**Comment by avkrash** [comment]
@bennieswart can you add @christian-bromann as reviewer, maybe it will hasten this PR

**Comment by pmerwin** [comment]
Awesome feature!  Thanks for working on it!

**Comment by bennieswart** [comment]
@christian-bromann I was waiting for your approval on the approach used before I went further. Will address those two points as soon as I get some free time.

**Comment by pmerwin** [comment]
@bennieswart  I have time to help, and would love to.

**Comment by bennieswart** [comment]
@pmerwin by all means go ahead!

**Comment by pmerwin** [comment]
@bennieswart 
Ok, for docs, we want to add to:
http://webdriver.io/guide/testrunner/retry.html ( I assume)
and a mocha test in tests for the test.

This will be my first time contributing, so I am not certain of what to do?  Can you point me in the direction of how I edit:

http://webdriver.io/guide/testrunner/retry.html

**Comment by bellstrand** [comment]
@pmerwin You can edit the docs in this repo. The file you're after is located at "[/docs/guide/testrunner/retry.md](https://github.com/webdriverio/webdriverio/blob/master/docs/guide/testrunner/retry.md)".

**Comment by pmerwin** [comment]
@bellstrand  ok will do!

**Comment by pdirt** [comment]
@bellstrand @bennieswart 

Here is the PR for the docs for this PR.  Please let me know if I have the info correct.

https://github.com/webdriverio/webdriverio/pull/2614

**Comment by bennieswart** [comment]
@christian-bromann I have added tests as requested.
It was somewhat tricky to determine whether or not the specfile-level retries are working, since each retry is a new process and hence state cannot be preserved. You're welcome to suggest a better approach than the current one if required.

**Comment by pdirt** [comment]
@bennieswart Is there any indication that a spec file failed and was retried in the logs?

File: lib/utils/ReporterStats.js [hunk 1]
```diff
@@ -3,14 +3,14 @@ import crypto from 'crypto'
 import sanitize from '../helpers/sanitize'
 
 class RunnableStats {
-    constructor (type) {
+    constructor (type, start) {
```
Comment (christian-bromann) [inline lib/utils/ReporterStats.js]: Can you elaborate why these changes are necessary?

File: test/fixtures/specs/retries.spec.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+console.log('Retry file loaded.')
+
+describe('retry', function () {
+    this.retries(2)
```
Comment (christian-bromann) [inline test/fixtures/specs/retries.spec.js]: Why are you including Mocha timeouts here?

File: test/spec/wdio/retries.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { spawn } from 'child_process'
+import path from 'path'
+import Launcher from '../../../build/lib/launcher'
+
+const STANDALONE_FLAG = '--run-launcher-standalone'
+
+if (process.argv.includes(STANDALONE_FLAG)) {
```
Comment (christian-bromann) [inline test/spec/wdio/retries.js]: What is the standalone doing?

File: lib/utils/ReporterStats.js (hunk 1 above)
Comment (bennieswart) [reply lib/utils/ReporterStats.js]: This is explained in a [previous comment](https://github.com/webdriverio/webdriverio/pull/2556#issuecomment-359717793) and has to do with incorrect timestamps of buffered messages.
> We've been running this patch for a couple of days now and the retries seem to be working nicely. I just noticed however that the reported duration of tests have dropped down to tens of millis. Only tests that have passed on their final run are correctly reporting duration. This correlates with the messages being buffered. I suspect start/end timestamps are assigned by the message handler as apposed to being contained within the message itself. Will look into this a bit later - should be an easy fix.

File: test/fixtures/specs/retries.spec.js (hunk 2 above)
Comment (bennieswart) [reply test/fixtures/specs/retries.spec.js]: I couldn't find any tests that tested the existing retry mechanisms, so I added some for those as well. I can remove them if you don't think they're necessary. I thought it couldn't hurt to have tests for retries on all three levels (test, suite, file).

Comment (christian-bromann) [reply test/fixtures/specs/retries.spec.js]: Got it but this should not be tested together with the spec retry. Having separate tests for it would make things easier I guess

File: test/spec/wdio/retries.js (hunk 3 above)
Comment (bennieswart) [reply test/spec/wdio/retries.js]: I'll address your overall comment here as well. The temporary file approach you suggest was something I considered as opposed to spawning a separate process for the retry-tests and checking the output. The reason I went with the separate process is that it doesn't require modifying the filesystem, whereas the temporary file obviously does. It seemed that, since the tests already spawn child processes, the separate process approach would be guaranteed to always work, whereas there could be situations in which modifying the filesystem would not (readonly access, for whatever reason).

Here, a separate process is spawned that runs the retry-tests and checks the output of that process to determine whether or not this test should pass or fail.
Depending on whether or not the flag is set, this file either *spawns* the new process, or *is* the new process. This could just as easily have been done using two different files, but keeping everything together made sense.

If you prefer the temporary file approach, I'll change it, no problem.

File: test/fixtures/specs/retries.spec.js (hunk 2 above)
Comment (bennieswart) [reply test/fixtures/specs/retries.spec.js]: Alright. Should I move them directly into `test/spec/wdio/retries.js` (since testing test-level and suite-level retries are quite trivial), or do you want them as a separate spec in the fixtures? Or something else?

**Comment by bennieswart** [comment]
@christian-bromann what are your thoughts on my previous comments? I'll make those changes once we're certain about the direction this is going.

**Comment by christian-bromann** [comment]
@bennieswart sorry for the late response, I will look into this asap

File: lib/launcher.js [hunk 4]
```diff
@@ -360,24 +355,51 @@ class Launcher {
...
+                cwd: process.cwd(),
+                execArgv
+            })
+
+            this.processes.splice(idx === -1 ? this.processes.length : idx, 1, childProcess)
+
+            childProcess.on('message', message => {
+                message._timestamp = new Date()
+                if (retries === 0) {
+                    this.messageHandler(cid, message)
```
Comment (christian-bromann) [inline lib/launcher.js]: Just got an idea from the Gitter channel, let's add the retry count to the message object so we can propagate the info to the reporters.

Comment (bennieswart) [reply lib/launcher.js]: The way things currently work will only ever send messages to the reporters when `retries === 0`.
When `retries > 0` and the specfile fails, buffered messages are discarded, and the reporters would be unaware that the failing testrun ever existed in the first place.

The reason for this is so that existing code continues to work. I suspect that if we suddenly report failing tests to reporters (even if they are retried soon afterwards) then a lot of things in the field will probably break.

Comment (christian-bromann) [reply lib/launcher.js]: > then a lot of things in the field will probably break.

Good point, let's leave stuff as is then

Comment (bennieswart) [reply lib/launcher.js]: I'm all for the idea of reporting the failures and retries to reporters, as this would allow for more insight into what happened. I just suspect things would become slightly more involved then than they currently are.
Mocha's suite-level and test-level retries are transparent (to my knowledge). Perhaps the specfile-level retries should be too.
Your call, ultimately.

**Comment by bennieswart** [comment]
@christian-bromann please provide feedback on the previous discussions so I know which direction to take the changes you have requested.
If you could list the outstanding changes along with your preference on each that would be great!

**Comment by dannsam** [comment]
@christian-bromann @bennieswart can I be of any help with this PR? Are there any outstanding comments that need to be addressed ?

**Comment by YmerejRedienhcs** [comment]
@bennieswart @christian-bromann any movement on this?  I seems like it would be very useful.

**Comment by caccialdo** [comment]
This is definitely a useful feature for us too. For now, we've forked this branch and use it as a private package but it would be great to see it officially part of the runner 👍

**Comment by bennieswart** [comment]
@christian-bromann are we still moving forward with this branch? I see the commit history has gone wacky (I haven't changed anything), but I can easily fix it if you still plan on moving forward.

**Comment by christian-bromann** [comment]
@bennieswart we updated the code basis to v5. Would you mind applying the patch to the new code basis?

**Comment by bennieswart** [comment]
@christian-bromann with regards to the discussion here https://github.com/webdriverio/webdriverio/pull/2556#discussion_r172349719, do you want notify reporters of the spec failure even though they are going to be retried? Or do you want to keep the failure transparent as is currently being done.

As mentioned previously, reporting failures (that will be retried) will probably break some existing things, but conversely the additional detail in the report may be useful. If ever there was a time to make such a (possibly) breaking change, I guess a major version bump is that time. Your call.

**Comment by christian-bromann** [comment]
> Or do you want to keep the failure transparent as is currently being done.

I think we should keep it transparent. It would be good if we would pass the retry count to reporters so they can use it to display how many retries the test took.

**Comment by bennieswart** [comment]
@christian-bromann I've had a look at the v5 codebase and it seems that an implementation for v5 would look quite different than the v4 implementation. Would it be possible to still merge this into v4 and then proceed to port it to v5 also? I'm guessing many people will be staying on v4 for quite some time to come until v5 has stabilized sufficiently (including me).

I've updated the source and destination branches of this PR, but the commit history is still broken. Perhaps it will also update when the PR is reopened. Otherwise, I'll just open a new one. That's if you're good with still getting this into v4.

**Comment by christian-bromann** [comment]
> I'll just open a new one.

Please open a new RR. Happy to merge it into v4.

**Comment by YmerejRedienhcs** [comment]
Thank you to both of you for this.
